cmd/query/app/querysvc/metrics_query_service.go

```diff
@@ -28,44 +28,47 @@ type MetricsQueryService struct {
 	metricsReader metricsstore.Reader
 }
 
-var errNilReader = errors.New("no reader defined for MetricsQueryService")
+var errMetricsQueryDisabled = errors.New("metrics querying is currently disabled")
 
 // NewMetricsQueryService returns a new MetricsQueryService.
-// A nil reader will result in a nil MetricsQueryService being returned.
 func NewMetricsQueryService(reader metricsstore.Reader) *MetricsQueryService {
 	return &MetricsQueryService{
 		metricsReader: reader,
 	}
 }
 
+func (mqs MetricsQueryService) Enabled() bool {
```

Comment:
Hopefully I understood your suggestion correctly, @yurishkuro.

Reply:
no, what I meant was adding this:
```
type nullMetricsQueryService struct {}
func (nullMetricsQueryService) GetLatencies(ctx context.Context, params *metricsstore.LatenciesQueryParameters) (*metrics.MetricFamily, error) { return nil, errNotEnabled }
...
```
and using it in RPC handlers. Basically, solve the problem with polymorphism, not with a bunch of nil checks all over the place.

Reply:
that's pretty cool :) although, unless if I'm mistaken, I see two minor issues:
- The handlers will need to parse inputs, etc. then call `GetLatencies`, before it knows that Metrics Querying is disabled. I think it would be better to return early; but not sure it's possible with this approach.
- Once calling `GetLatencies`, the handler needs to differentiate the `errNotEnabled` from an "exceptional" error case so it can return an appropriate error code rather than the generic `StatusInternalServerError` code, which is important as the UI needs to handle both cases differently. 
  We could follow a pattern like [this](https://github.com/jaegertracing/jaeger/blob/master/cmd/query/app/http_handler.go#L415) but there's no guarantee the error could be wrapped up, so I imagine we'd need use something like `errors.Is`.

what do you think?

Reply:
and yes, you'd want to use `error.Is`, it's the new Go way, what we have with `if err == spanstore.ErrTraceNotFound {` predates that.

Reply:
> what's the downside of parsing inputs?

not a large downside, and the emphasis wasn't so much on the input parsing but on when the handler knows if the feature is enabled or disabled. I feel the intent is clearer when reading the handler code instead of needing to follow it to the point where the error is handled:
```
metricsHandler() {
    if notEnabled {
       return notImplementedErrorCode
    }
    res, err := service.call()
    ...
}
```

In any case, it's a small matter I think, and happy to go ahead with your suggestion.